sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/PooledHashMap.java

```diff
@@ -0,0 +1,253 @@
+   * buckets expands over time to avoid having too many items in one bucket, otherwise accessing an
+   * item by key won't be a constant time complexity.
+   *
+   * @param capacity The initial number of buckets to start with
+   */
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  public PooledHashMap(int capacity) {
+    this.table = new ArrayList[capacity];
+    this.entryPool = new ObjectPool<>(Entry::new);
+    this.size = 0;
+  }
+
+  /**
+   * Creates a new {@link PooledHashMap} with a default amount of buckets (capacity).
+   *
+   * @see PooledHashMap#PooledHashMap(int)
+   */
+  public PooledHashMap() {
+    this(DEFAULT_CAPACITY);
+  }
+
+  /**
+   * Add a key, value pair to the map.
+   *
+   * <p>Internally it uses a MapEntry from a pool of entries, to store this mapping
+   *
+   * @param key key with which the specified value is to be associated
+   * @param value value to be associated with the specified key
+   * @return Null if the was no previous mapping for this key, or the value of the previous mapping
+   *     of this key
+   */
+  @Override
+  @Nullable
+  public V put(K key, V value) {
+    Objects.requireNonNull(key, "This map does not support null keys");
+    Objects.requireNonNull(value, "This map does not support null values");
+    if (size > LOAD_FACTOR * table.length) {
+      rehash();
+    }
+
+    int bucket = getBucket(key);
+    ArrayList<Entry<K, V>> entries = table[bucket];
+    if (entries == null) {
+      entries = new ArrayList<>();
+      table[bucket] = entries;
+    } else {
+      for (int i = 0; i < entries.size(); i++) {
+        Entry<K, V> entry = entries.get(i);
+        if (Objects.equals(entry.key, key)) {
+          V oldValue = entry.value;
+          entry.value = value;
+          return oldValue;
+        }
+      }
+    }
+    Entry<K, V> entry = entryPool.borrowObject();
+    entry.key = key;
+    entry.value = value;
+    entries.add(entry);
+    size++;
+    return null;
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked", "NullAway"})
+  private void rehash() {
+    ArrayList<Entry<K, V>>[] oldTable = table;
+    table = new ArrayList[2 * oldTable.length];
+
+    // put() to new table below will reset size back to correct number
+    size = 0;
+
+    for (int i = 0; i < oldTable.length; i++) {
+      ArrayList<Entry<K, V>> bucket = oldTable[i];
+      if (bucket != null) {
+        bucket.forEach(
+            entry -> {
+              put(entry.key, entry.value);
+              entryPool.returnObject(entry);
+            });
+        bucket.clear();
+      }
+    }
+  }
+
+  /**
+   * Retrieves the mapped value for {@code key}.
+   *
+   * @param key the key whose associated value is to be returned
+   * @return The mapped value for {@code key} or null if there is no such mapping
+   */
+  @Override
+  @Nullable
+  @SuppressWarnings("unchecked")
+  public V get(Object key) {
+    int bucket = getBucket((K) key);
```

Comment:
Can we add null checks all methods like in `put`?
```
Objects.requireNonNull(key, "This map does not support null keys");
```

Reply:
Looks like we're still missing a null check in `remove` and `containsKey`

Reply:
Thanks. Added.